packages/liveblocks-react-ui/src/components/AiChat.tsx

```diff
@@ -133,49 +414,63 @@ export const AiChat = forwardRef<HTMLDivElement, AiChatProps>(
 
     const containerRef = useRef<HTMLDivElement | null>(null);
-    const containerBottomRef = useRef<HTMLDivElement | null>(null);
-    const isScrollIndicatorEnabled = !isLoading && !error;
-    const isScrollAtBottom = useVisible(containerBottomRef, {
-      enabled: isScrollIndicatorEnabled,
-      root: containerRef,
-      rootMargin: MIN_DISTANCE_BOTTOM_SCROLL_INDICATOR,
-      initialValue: null,
-    });
-    const isScrollIndicatorVisible =
-      isScrollIndicatorEnabled && isScrollAtBottom !== null
-        ? !isScrollAtBottom
-        : false;
+    const messagesRef = useRef<HTMLDivElement | null>(null);
+    const footerRef = useRef<HTMLDivElement | null>(null);
+    const scrollBottomRef = useRef<HTMLDivElement | null>(null);
+    const trailingSpacerRef = useRef<HTMLDivElement | null>(null);
 
-    const [lastSentMessageId, setLastSentMessageId] =
-      useState<MessageId | null>(null);
+    const [isScrollAtBottom, setScrollAtBottom] = useState<boolean | null>(
+      null
+    );
+    // `useState`'s setter is stable but this is for clarity in the places it's used.
+    const onScrollAtBottomChange = useLatest(setScrollAtBottom);
+    const isScrollIndicatorVisible =
+      messages && isScrollAtBottom !== null ? !isScrollAtBottom : false;
 
     useImperativeHandle<HTMLDivElement | null, HTMLDivElement | null>(
       forwardedRef,
       () => containerRef.current,
       []
     );
 
-    const scrollToBottomCallbackRef =
-      useRef<(behavior: "instant" | "smooth") => void>(undefined);
-    if (scrollToBottomCallbackRef.current === undefined) {
-      scrollToBottomCallbackRef.current = function (
-        behavior: "instant" | "smooth"
-      ) {
+    const scrollToBottom = useLatest(
+      (behavior: "instant" | "smooth", includeTrailingSpace = false) => {
         const container = containerRef.current;
-        if (container === null) return;
 
-        container.scrollTo({
-          top: container.scrollHeight,
-          behavior,
-        });
-      };
-    }
-    const scrollToBottom = scrollToBottomCallbackRef.current;
+        if (!container) {
+          return;
+        }
+
+        if (includeTrailingSpace) {
+          // Scroll all the way to include the trailing space,
+          // and wait for a frame in case the trailing space hasn't
+          // been updated yet. (e.g. when sending a new message)
+          requestAnimationFrame(() => {
+            container.scrollTo({
+              top: container.scrollHeight,
+              behavior,
+            });
+          });
+        } else {
+          const messagesHeight = messagesRef.current?.offsetHeight ?? 0;
+          const footerHeight = footerRef?.current?.offsetHeight ?? 0;
+
+          // Scroll to the bottom of the messages to exclude the trailing space.
+          container.scrollTo({
+            top: messagesHeight + footerHeight - container.clientHeight,
+            behavior,
+          });
```

Comment:
I see - `includeTrailingSpace` is `true` when a new user message is sent.

In some of my past experiments, I've tried doing something around the lines of (because we already have those elements ready to be scrolled into view):

```ts
if (includeTrailingSpace) {
  footerRef.current?.scrollIntoView({ behavior: "smooth", block: "end", inline: "nearest" })
} else {
  scrollBottomRef.current?.scrollIntoView({ behavior: "smooth" })
}
```

Have you considered something like this?

Reply:
I tried it in the past but on the messages list and I wasn't satisfied with the lack of control, but now that we have two 0px "`IntersectionObserver` markers" (one before the trailing space, one after) you're right we can definitely avoid the manual measurement with `scrollTo`, done in https://github.com/liveblocks/liveblocks/pull/2520/commits/4250c2adbd6c8fe42d14647be0e98ef4acf598b2.